homeassistant/components/event/__init__.py

```diff
@@ -0,0 +1,185 @@
+    _attr_event_types: list[str]
+    _attr_state: None
+    _attr_extra_state_attributes: None  # type: ignore[assignment]
+
+    __last_event: datetime | None = None
+    __last_event_type: str | None = None
+    __last_event_extra_state_attributes: dict[str, Any] | None = None
+
+    @property
+    def device_class(self) -> EventDeviceClass | None:
+        """Return the class of this entity."""
+        if hasattr(self, "_attr_device_class"):
+            return self._attr_device_class
+        if hasattr(self, "entity_description"):
+            return self.entity_description.device_class
+        return None
+
+    @property
+    def event_types(self) -> list[str]:
+        """Return a list of possible events."""
+        if hasattr(self, "_attr_event_types"):
+            return self._attr_event_types
+        if (
+            hasattr(self, "entity_description")
+            and self.entity_description.event_types is not None
+        ):
+            return self.entity_description.event_types
+        raise AttributeError()
+
+    @final
+    @callback
+    def async_trigger_event(
+        self, event_type: str, extra_state_attributes: dict[str, Any] | None = None
+    ) -> None:
+        """Process a new event."""
+        if event_type not in self.event_types:
+            raise ValueError(f"Invalid event type {event_type} for {self.entity_id}")
+        self.__last_event = dt_util.utcnow()
+        self.__last_event_type = event_type
+        self.__last_event_extra_state_attributes = extra_state_attributes
+        self.async_write_ha_state()
+
+    @final
+    def trigger_event(
+        self, event_type: str, extra_state_attributes: dict[str, Any] | None = None
+    ) -> None:
+        """Process a new event."""
+        self.hass.loop.call_soon_threadsafe(
+            self.async_trigger_event, event_type, extra_state_attributes
+        )
+
+    def _default_to_device_class_name(self) -> bool:
+        """Return True if an unnamed entity should be named by its device class.
+
+        For events this is True if the entity has a device class.
+        """
+        return self.device_class is not None
+
+    @property
+    @final
+    def capability_attributes(self) -> dict[str, list[str]]:
+        """Return capability attributes."""
+        return {
+            ATTR_EVENT_TYPES: self.event_types,
+        }
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the entity state."""
+        if self.__last_event is None:
+            return None
+        return self.__last_event.isoformat(timespec="milliseconds")
+
+    @final
+    @property
+    def state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes."""
+        return {ATTR_EVENT_TYPE: self.__last_event_type}
+
+    @property
+    @final
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return entity specific state attributes."""
+        return self.__last_event_extra_state_attributes
```

Comment:
MQTT overrides this attribute, is it needed to mark this @final ?

Reply:
Yes, in this proposal it is, as it is protected and managed (including restore) by the entity. In this specific version it becomes an internal.

I am finishing up a second alternative that is more "old-fashioned", so we have something to discuss.

Reply:
I've tried the second approach and came to the conclusion that will be much worse.